modules/runner/src/main/ts/core/Reporter.ts

```diff
@@ -0,0 +1,163 @@
+import { LoggedError, Reporter } from '@ephox/bedrock-common';
+import { Callbacks, onSuccessCallback } from './Callbacks';
+import { UrlParams } from './UrlParams';
+import { formatElapsedTime, makeUrl } from './Utils';
+
+export interface TestReporter {
+  start: (onDone: onSuccessCallback) => void;
+  pass: (onDone: onSuccessCallback) => void;
+  fail: (e: LoggedError.LoggedError, onDone: onSuccessCallback) => void;
+}
+
+export interface Reporter {
+  summary: () => { passed: number; failed: number };
+  test: (file: string, name: string) => TestReporter;
+  done: () => void;
+  setStopOnFailure: (stopOnFailure: boolean) => void;
+  shouldStopOnFailure: () => boolean;
+}
+
+declare const $: JQueryStatic;
+
+const elapsed = (since: Date): string => formatElapsedTime(since, new Date());
+
+export const dom = (container: JQuery<HTMLElement>, totalNumTests: number, params: UrlParams, callbacks: Callbacks): Reporter => {
+  const initial = new Date();
+  let passCount = 0;
+  let failCount = 0;
+  let stopOnFailure = false;
+
+  const summary = () => ({
+    passed: passCount + (params.offset - params.failed),
+    failed: failCount + params.failed,
+  });
+
+  const current = $('<span />').addClass('progress').text(params.offset);
+  const restartBtn = $('<button />').text('Restart').on('click', () => {
+    const url = makeUrl('', 0, 0, 0);
+    window.location.assign(url);
+  });
+  const retryBtn = $('<button />').text('Retry').on('click', () => {
```

Comment:
Minor: it seems that all of these functions are the same. 

1. Calculate a summary only if required by the makeUrl command
2. Call makeUrl with either the values from the summary, or zeros and empty strings
3. Reassign the page URL.

If you were doing the entirely unimportant change mentioned above to pass in the handlers etc. you could pass in the same handler for all of them, and just pass that handler the arguments to the `makeUrl` function here. Or could pass a type to the handler, and it interpreted that type (retry test, skip test, restart tests etc.)

Just a thought.

Reply:
Done, though in a bit of a different way...